cluster-autoscaler/cloudprovider/kwok/kwok_provider.go

```diff
@@ -0,0 +1,231 @@
+func (kwok *KwokCloudProvider) GetResourceLimiter() (*cloudprovider.ResourceLimiter, error) {
+	return kwok.resourceLimiter, nil
+}
+
+// GPULabel returns the label added to nodes with GPU resource.
+func (kwok *KwokCloudProvider) GPULabel() string {
+	// GPULabel() might get called before the config is loaded
+	if kwok.config == nil || kwok.config.status == nil {
+		return ""
+	}
+	return kwok.config.status.gpuLabel
+}
+
+// GetAvailableGPUTypes return all available GPU types cloud provider supports
+func (kwok *KwokCloudProvider) GetAvailableGPUTypes() map[string]struct{} {
+	// GetAvailableGPUTypes() might get called before the config is loaded
+	if kwok.config == nil || kwok.config.status == nil {
+		return map[string]struct{}{}
+	}
+	return kwok.config.status.availableGPUTypes
+}
+
+// TODO(vadasambar): check how to make the response from this non-nil (it's nil right now)
+// GetNodeGpuConfig returns the label, type and resource name for the GPU added to node. If node doesn't have
+// any GPUs, it returns nil.
+func (kwok *KwokCloudProvider) GetNodeGpuConfig(node *apiv1.Node) *cloudprovider.GpuConfig {
+	return gpu.GetNodeGPUFromCloudProvider(kwok, node)
+}
+
+// TODO(vadasambar): implement this
+// Refresh is called before every main loop and can be used to dynamically update cloud provider state.
+// In particular the list of node groups returned by NodeGroups can change as a result of CloudProvider.Refresh().
+func (kwok *KwokCloudProvider) Refresh() error {
+
+	// TODO(vadasambar): causes CA to not recognize kwok nodegroups
+	// needs better implementation
+	// nodeList, err := kwok.lister.List()
+	// if err != nil {
+	// 	return err
+	// }
+
+	// ngs := []*NodeGroup{}
+	// for _, no := range nodeList {
+	// 	ng := parseAnnotationsToNodegroup(no)
+	// 	ng.kubeClient = kwok.kubeClient
+	// 	ngs = append(ngs, ng)
+	// }
+
+	// kwok.nodeGroups = ngs
+
+	return nil
+}
+
+// Cleanup cleans up all resources before the cloud provider is removed
+func (kwok *KwokCloudProvider) Cleanup() error {
+	nodeList, err := kwok.lister.List()
+	if err != nil {
+		return fmt.Errorf("error cleaning up kwok nodes: %v", err)
+	}
+	for _, node := range nodeList {
+		err := kwok.kubeClient.CoreV1().Nodes().Delete(context.Background(), node.GetName(), v1.DeleteOptions{})
+		if err != nil {
+			klog.Errorf("error cleaning up node '%v'", node.GetName())
+		}
+	}
+
+	return nil
+}
+
+// BuildKwok builds kwok cloud provider.
+func BuildKwok(opts config.AutoscalingOptions, do cloudprovider.NodeGroupDiscoveryOptions, rl *cloudprovider.ResourceLimiter) cloudprovider.CloudProvider {
+	restConfig, err := rest.InClusterConfig()
+	if err != nil {
+		klog.Fatalf("failed to get kubeclient config for external cluster: %v", err)
+	}
+
+	kubeClient := kubeclient.NewForConfigOrDie(restConfig)
+
+	stopCh := make(chan struct{})
+	lister := kube_util.NewNodeLister(kubeClient, func(n *apiv1.Node) bool {
+		return n.GetAnnotations()[KwokManagedAnnotation] == "fake"
+	}, stopCh)
```

Comment:
These code is out of date. CA can run out of cluster and the source of rest config should be same with kube client used in main.go:

https://github.com/kubernetes/autoscaler/blob/bafcc0dfaf887975cba31a20f81cd47dddb2a06f/cluster-autoscaler/main.go#L444

I opened a new PR to refactor those code expecting to reuse the rest config: https://github.com/kubernetes/autoscaler/pull/6180